### PR TITLE
easylogging++: cached allowed categories

### PR DESCRIPTION
**Pull request by moneromooo-monero**
It turns out this can be fairly heavy when logging a lot

### CODE REVIEW

File: external/easylogging++/easylogging++.cc [hunk 1]
```diff
@@ -2033,15 +2035,21 @@ static int priority(Level level) {
...
   base::threading::ScopedLock scopedLock(lock());
+  const std::map<std::pair<std::string, Level>, bool>::const_iterator it = m_cached_allowed_categories.find(std::make_pair(category, level));
+  if (it != m_cached_allowed_categories.end())
+    return it->second;
   if (m_categories.empty() || category == nullptr) {
     return false;
   } else {
     std::deque<std::pair<std::string, Level>>::const_reverse_iterator it = m_categories.rbegin();
     for (; it != m_categories.rend(); ++it) {
       if (base::utils::Str::wildCardMatch(category, it->first.c_str())) {
```
Comment (vtnerd) [inline external/easylogging++/easylogging++.cc]: There is still a lot of cycles here just to determine whether something should be logged: (1) A lock must be acquired, (2) a `std::string` must be allocated (for the `std::map` call _until_ C++14 templated `std::map::find`), (3) _possibly_ walk this list + allocate another `std::string` again.

Is the wildcard matching useful? The `std::deque` could be a `std::unordered_map` and then no caching would be needed. Or even a middle ground - the `std::deque` could be sorted, and then a `std::lower_bound` / `std::upper_bound` (ranged) search can still be done on the prefix before `*` or `?` characters.

At a minimum, I would suggest making this a `std::map<std::string, Level>` or `std::unordered_map<std::string, Level>`. Its going to reduce the number of entries needed, and a temporary `std::string` can be re-used (and even moved) from find to insert calls. Or just use `category` directly 3 times (like now), and make it C++14 ready.

Comment (vtnerd) [reply external/easylogging++/easylogging++.cc]: Crap, I got the wildcard pattern reversed - `category` is never wildcarded, only the map entry. There are actually two cases then, a map entry with a prefix, and one without. So a ranged search is still possible, but you would have to disallow non-prefixed categories or walk through them separately.

Comment (moneromooo-monero) [reply external/easylogging++/easylogging++.cc]: Wildcards are useful.
If I use a map<string,Level>, then I think I'd need an extra fake level which is "nothing" to cache the case where a category is not found in the list.

Comment (moneromooo-monero) [reply external/easylogging++/easylogging++.cc]: Actually I can just use ints since priority(Level) returns those, I'l try that.

Comment (vtnerd) [reply external/easylogging++/easylogging++.cc]: The source of the problem is that categories are strings. There are a limited number of categories that can come from the code, and therefore it should be based on an enumerated type. When someone sets/updates the log level string, the wildcard logic can be executed once against the fixed set of categories. When logging, this check becomes a constant operation, and the lock can even be removed by using atomics.

The upstream source doesn't have the category feature, so this source has already divered from upstream.

**Comment by vtnerd** [comment]
This looks fine. Although we need to strongly consider moving to an enumerated categories system - the lock can be removed entirely for an array of atomics based on enumerated value. The advantage is that a call to log only locks up the cacheline when something is actually written (currently it is a bottleneck for all calls to the logger).

**Pull request merged**